counterexamples/zero_divisors_in_add_monoid_algebras.lean

```diff
@@ -0,0 +1,211 @@
+        (nsmul_ne_zero_of_lt_add_order_of' one_ne_zero (nat.succ_le_iff.mp o2)),
+      simp only [a0, single_eq_of_ne, ne.def, not_false_iff] },
+    { simpa only [single_eq_same] using zero_ne_one, } },
+  { convert commute.geom_sum₂_mul _ (add_order_of a),
+    { ext, rw [single_zero_one, one_pow, mul_one] },
+    { rw [single_pow, one_pow, add_order_of_nsmul_eq_zero, single_zero_one, one_pow, sub_self] },
+    { simp only [single_zero_one, commute.one_right] } },
+end
+
+example {R} [ring R] [nontrivial R] (n : ℕ) (n0 : 2 ≤ n) :
+  ∃ f g : add_monoid_algebra R (zmod n), f ≠ 0 ∧ g ≠ 0 ∧ f * g = 0 :=
+zero_divisors_of_torsion (1 : zmod n) (n0.trans_eq (zmod.add_order_of_one _).symm)
+
+/--  `F` is the type with two elements `zero` and `one`.  We define the "obvious" linear order and
+absorbing addition on it to generate our counterexample. -/
+@[derive [decidable_eq, inhabited]] inductive F | zero | one
+
+/--  The same as `list.get_rest`, except that we take the "rest" from the first match, rather than
+from the beginning, returning `[]` if there is no match.  For instance,
+```lean
+#eval [1,2].drop_until [3,1,2,4,1,2]  -- [4, 1, 2]
+```
+-/
+def list.drop_until {α} [decidable_eq α] : list α → list α → list α
+| l [] := []
+| l (a::as) := ((a::as).get_rest l).get_or_else (l.drop_until as)
+
+/-- `guard_decl_in_file na loc` makes sure that the declaration with name `na` is in the file with
+relative path `"src/" ++ "/".intercalate loc ++ ".lean"`.
+```lean
+#eval guard_decl_in_file `nat.nontrivial ["data", "nat", "basic"]  -- does nothing
+
+#eval guard_decl_in_file `nat.nontrivial ["not", "in", "here"]
+-- fails giving the location 'data/nat/basic.lean'
+```
+
+This test makes sure that the comment referring to this example is in the file claimed in the
+doc-module to this counterexample. -/
+meta def guard_decl_in_file (na : name) (loc : list string) : tactic unit :=
+do env ← tactic.get_env,
+  some fil ← pure $ env.decl_olean na | fail!"the instance `{na}` is not imported!",
+  let path : string := ⟨list.drop_until "/src/".to_list fil.to_list⟩,
+  let locdot : string := ".".intercalate loc,
+  guard (fil.ends_with ("src/" ++ "/".intercalate loc ++ ".lean")) <|>
+    fail!("instance `{na}` is no longer in `{locdot}`.\n\n" ++
+      "Please, update the doc-module and this check with the correct location:\n\n'{path}'\n")
+
+#eval guard_decl_in_file `finsupp.lex.covariant_class_le_left ["data", "finsupp", "lex"]
+#eval guard_decl_in_file `finsupp.lex.covariant_class_le_right ["data", "finsupp", "lex"]
+
+namespace F
+
+instance : has_zero F := ⟨F.zero⟩
+
+/--  `1` is not really needed, but it is nice to use the notation. -/
+instance : has_one F := ⟨F.one⟩
+
+/--  A tactic to prove trivial goals by enumeration. -/
+meta def boom : tactic unit :=
+`[ repeat { rintro ⟨⟩ }; dec_trivial ]
+
+/--  `aux` maps `0 1 : F` to their counterparts in `ℕ`.
+We use it to lift the linear order on `N`. -/
+def aux : ∀ a : F, ℕ
```

Comment:
A slightly better name might be:
```suggestion
def val : ∀ a : F, ℕ
```

Reply:
Oh, that works too!  Not sure why I had this weird form to introduce the pattern-matching!